emily/lambda/src/operations/chainstate.rs

```diff
@@ -2,50 +2,148 @@
 
 use std::collections::HashMap;
 
+use aws_sdk_dynamodb::types::AttributeValue;
 use emily::models;
 use crate::common;
-use crate::errors;
+use crate::config::LambdaContext;
+use crate::errors::{self, EmilyApiError};
 
 /// Handles the creation of a new chainstate
-pub fn handle_create_chainstate(
-    body:  Option<String>
+pub async fn handle_create_chainstate(
+    body: Option<String>,
+    context: &LambdaContext,
 ) -> Result<common::SimpleApiResponse, errors::EmilyApiError> {
-    common::deserialize_request::<models::CreateChainstateRequestContent>(body).map(|request| models::CreateChainstateResponseContent {
-                block_hash: request.block_hash,
-                block_height: request.block_height,
-            })
-        .and_then(|response| {
-            common::package_response(response, 201)
-        })
+
+    let request: models::CreateChainstateRequestContent =
+        common::deserialize_request::<models::CreateChainstateRequestContent>(body)?;
+
+    let chainstate_entry = models::CreateChainstateResponseContent {
+        block_hash: request.block_hash.clone(),
+        block_height: request.block_height,
+    };
+
+    let mut item: HashMap<String, AttributeValue> = HashMap::new();
+    item.insert("BlockHeight".to_string(), AttributeValue::N(request.block_height.to_string()));
+    item.insert("BlockHash".to_string(), AttributeValue::S(request.block_hash));
+
+    context.dynamodb_client
+        .put_item()
+        .table_name(context.settings.chainstate_table_name.clone())
+        .set_item(Some(item))
```

Comment:
We might want to add a [condition expression](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Expressions.ConditionExpressions.html#Expressions.ConditionExpressions.PreventingOverwrites) for whether the block exists by doing something like
```rust
    ...
    .put_item()
    .set_item(Some(item))
    .condition_expression("attribute_not_exists(block_hash)")
    ...
```

Addendum: Well, this only makes sense if this table has a timestamp column, since that is the only thing that is getting updated when we overwrite an entry that already exists. If timestamps are left out here, no need to include it.

Reply:
Yeah nevermind, this isn't needed.

Reply:
> So setting an item in dynamodb isn't idempotent, but will fail if the item already exists? I didn't know that. Good point!

Well, setting items in dynamodb is indeed idempotent. But if we add a timestamp attribute then the timestamp would likely change if the caller hits the endpoint twice. If we don't want to update that attribute then we'd have to add a condition expression.

Reply:
Right, makes sense!

Reply:
Instead of using a timestamp for this I'm using a "version" field that will increment whenever there's an update; I think this is the more standard approach for dynamodb.

Reply:
> Instead of using a timestamp for this I'm using a "version" field that will increment whenever there's an update; I think this is the more standard approach for dynamodb.

Ahhh I see. And that's why you want a PUT and a POST, you only need to do a get for the `version` with a PUT.